Idea: Provide initial data to the StreamBuilder

brianegan opened the pull request:
This allows one to run the builder function the first time with an `AsyncSnapshot` that contains the given `initialData`, rather than with an `AsyncSnapshot` that has no data.

This can be helpful when the Data your Stream provides is itself a State Machine with it's own "has data" / "missing data" semantics.

As it currently works, you need to deal with not only the `AsyncSnapshot` State, but also your own Data's State, which leads to additional / clunky branching logic.

brianegan committed:
Provide initial data to the StreamBuilder

This allows one to run the builder function the first time with the
given initialData.

brianegan commented:
CLA has been signed

mravn-google reviewed (commented):
Thanks for contributing. Nice idea!

We should make sure that

1. A `StreamBuilder` made from a `Future` using `Future.asStream()` should behave in the same way as the corresponding `FutureBuilder`. This implies that we should add the initial data parameter to `FutureBuilder` also (API consistency).
1. A `StreamBuilder` or `FutureBuilder` with `null` initial data should behave identically to the current `StreamBuilder`/`FutureBuilder` (avoids breaking backwards compatibility).
1. A `StreamBuilder` or `FutureBuilder` with non-null initial data x should start in the same state as a builder that has 1) previously been connected to another stream or future, 2) received x from that source, then 3) connected instead to the current stream/future (avoids introducing new states).

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -333,19 +333,27 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
   /// Creates a new [StreamBuilder] that builds itself based on the latest
   /// snapshot of interaction with the specified [stream] and whose build
-  /// strategy is given by [builder].
+  /// strategy is given by [builder]. You can also specify the [initialData],
```
The semantics of `initialData` should be clarified, cf. the general comments to this PR.

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -333,19 +333,27 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
   /// snapshot of interaction with the specified [stream] and whose build
-  /// strategy is given by [builder].
+  /// strategy is given by [builder]. You can also specify the [initialData],
+  /// which will be contained within the first AsyncSnapshot delivered to
```
AsyncSnapshot -> snapshot

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -333,19 +333,27 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
-  /// strategy is given by [builder].
+  /// strategy is given by [builder]. You can also specify the [initialData],
+  /// which will be contained within the first AsyncSnapshot delivered to
+  /// your [builder].
```
your -> the

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -333,19 +333,27 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
   /// The build strategy currently used by this builder. Cannot be null.
   final AsyncWidgetBuilder<T> builder;
 
+  /// Allows you to provide data for the first AsyncSnapshot sent to the
```
AsyncSnapshot -> snapshot

mravn-google replied:
Comment should reflect semantics of `initialData` as per general comment to this PR.

brianegan committed:
Improve test coverage, docs, and api consistency

brianegan commented:
Thanks for the feedback, @mravn-google! I've updated the code address your points:

  1. Add `initialData` to `FutureBuilder` for API consistency. Add tests to cover `FutureBuilder` and API consistency.
  2. The `null` case is covered by existing tests, which were not affected by this change and continue to work.
  3. Added tests to ensure changing the `Future` or `Stream` does not result in new States.
  4. Tried to add clearer / more semantic documentation to closer fit in with the other comments, hope it fits with what you were suggesting. Feedback welcome of course!

One interesting case: What if you change the `Future` or `Stream` and *also* change the `initialData`. Currently, this does not result in any state changes. Should it? I don't think it should, but wanted to raise it as a question.

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -333,19 +343,25 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
   /// snapshot of interaction with the specified [stream] and whose build
-  /// strategy is given by [builder].
+  /// strategy is given by [builder]. The [initialData] is used to create the
+  /// first snapshot delivered to the [builder].
```
This sentence is incorrect. While `initialData` controls the initial value of `_snapshot`, that value is only given to `builder`, if the `build` method is invoked by the Flutter framework prior to the receipt of the first stream event.

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -391,6 +407,16 @@ class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
 /// * `new AsyncSnapshot<String>.withData(ConnectionState.waiting, null)`
 /// * `new AsyncSnapshot<String>.withError(ConnectionState.done, 'some error')`
 ///
+/// By default, the initial snapshot delivered to the [builder] will not contain
```
As above. Maybe say
```dart
/// The initial snapshot data can be controlled by specifying [initialData]. You would use
/// this facility to ensure that if the [builder] is invoked before the future completes, the
/// snapshot carries data of your choice rather than the default null value.
```

mravn-google commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -458,11 +488,12 @@ class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
   @override
   void initState() {
     super.initState();
+    _snapshot = new AsyncSnapshot<T>.withData(ConnectionState.none, widget.initialData); // ignore: prefer_const_constructors
```
Can't we remove the `ignore:` here?

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -66,6 +66,19 @@ void main() {
+      expect(find.text('AsyncSnapshot(ConnectionState.none, $initial, null)'), findsOneWidget);
+      final Completer<String> completer = new Completer<String>();
+      await tester.pumpWidget(new FutureBuilder<String>(
+        key: key, future: completer.future, builder: snapshotText, initialData: 'No change',
```
Maybe `'Ignored'` instead of `'No change'`. Similarly in other test cases below.

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -66,6 +66,19 @@ void main() {
       expect(find.text('AsyncSnapshot(ConnectionState.waiting, null, null)'), findsOneWidget);
     });
+    testWidgets('gracefully handles transition from null future with initial data without introducing new states', (WidgetTester tester) async {
+      final String initial = 'initial';
```
As a matter of personal preference (feel free to ignore), I find it easier to read a short test case that uses literal data rather than variables which have to be declared and interpolated into strings. Even if that literal has to be repeated once or twice. Some of the later test cases use `'A'` and `'B'` in this manner.

I'd probably just use `'I'` for the initial data.

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -81,6 +94,22 @@ void main() {
       await eventFiring(tester);
       expect(find.text('AsyncSnapshot(ConnectionState.none, null, null)'), findsOneWidget);
     });
+    testWidgets('gracefully handles transition to null future with initial data without introducing new states', (WidgetTester tester) async {
```
About the test name: What is actually tested is that when reconfiguring a `FutureBuilder`, the `initialData` is ignored. So I suggest renaming the test to `ignores initialData when reconfiguring`.

I think we should avoid retesting that the `FutureBuilder` gracefully handles its transitions. The `initialData` feature is orthogonal to that. That should cut down on the number of test cases below.

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -180,6 +273,19 @@ void main() {
       expect(find.text('AsyncSnapshot(ConnectionState.done, 4, null)'), findsOneWidget);
     });
+    testWidgets('runs the builder using given initial data', (WidgetTester tester) async {
+      final String initial = 'A';
```
Please use a different value than `'A'`. That value is associated with stream A or future A in other test cases.

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -211,6 +317,25 @@ void main() {
+      ]));
+      expect(find.text('AsyncSnapshot(ConnectionState.none, A, null)'), findsNWidgets(2));
+    });
+    testWidgets('when using initialData and completes with data', (WidgetTester tester) async {
```
completes -> completing

mravn-google reviewed (commented):
Thanks for your effort!

brianegan commented:
Cool, thanks again for the feedback! I'll push up another commit addressing those points.

brianegan commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -303,6 +303,16 @@ typedef Widget AsyncWidgetBuilder<T>(BuildContext context, AsyncSnapshot<T> snap
+/// By default, the initial snapshot will not contain data, resulting in a
+/// snapshot of the form
+///
+/// * `new AsyncSnapshot<String>.withData(ConnectionState.none, null)`
```
All docs updated to use [builder] rather than [build].

brianegan committed:
Improve code based on feedback

brianegan commented:
All right, tried to properly address all comments in the new commit.

  * Improve docs based on feedback
  * Removed graceful transition tests for `StreamBuilder` and `FutureBuilder`, replaced with two tests for each:
    1.  "runs the builder using given initial data"
    2. "ignores initialData when reconfiguring" -- Technically, the first test is also covered by this test, but kept them separate for clarity. I could also collapse them into one test case if you'd like.

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -211,6 +251,24 @@ void main() {
     });
+    testWidgets('when initialData is used with null Future and Stream', (WidgetTester tester) async {
+      await tester.pumpWidget(new Column(children: <Widget>[
+        new FutureBuilder<String>(future: null, builder: snapshotText, initialData: 'I',),
```
[nit] Avoid trailing comma for single line parameter lists.

brianegan replied:
Trailing commas removed.

mravn-google reviewed (approved):
LGTM, thanks!

brianegan committed:
Remove trailing commas

brianegan commented:
Yay! Thanks for helping me get this PR through :)

mravn-google commented on packages/flutter/test/widgets/async_test.dart:
```diff
@@ -118,6 +118,25 @@ void main() {
+    testWidgets('runs the builder using given initial data', (WidgetTester tester) async {
+      final GlobalKey key = new GlobalKey();
+      await tester.pumpWidget(new FutureBuilder<String>(
+        key: key, future: null, builder: snapshotText, initialData: 'I'
```
Here and similar places below, I think we'd want both one line per parameter and a trailing comma. Either we do
```dart
 aFunctionCallThatFitsOnOneLine(some: key, no: trailing, comma: see);
```
or we do
```dart
 aFunctionCallThatNeedsMultipleLines(
   because: we,
   have: so,
   many: parameters,
   andWeThenNeed: a,
   trailing: comma,
 );
```
The point of doing the above is that 1) it obeys the Flutter style guide, and 2) `dartfmt` leaves it unchanged.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line

brianegan replied:
Yah no worries... that's how I normally do it, but I wasn't 100% sure about the Flutter style b/c this file is full of lines like that:

https://github.com/flutter/flutter/pull/13820/files#diff-ecef29ade55b943f96ea50b77461443fR88

Should I file a follow-up issue to fix the styling for this file?

mravn-google replied:
I see your point. We'll leave it for now then. I believe Flutter style will eventually align with `dartfmt`, and we should generally stay as close to `dartfmt`'s output as possible without violating the Flutter style guide. I'll merge this PR.

Hixie replied:
This particular file isn't yet in strict adherence to either dartfmt or the Flutter style guide, unfortunately.

Hixie replied:
(I don't mean the changes from this PR, I mean the file as it was before this PR.)

mravn-google replied:
What's wrong? (Apart from an very long line towards the end?)

Hixie replied:
Oh, my bad. I had meant the same lines you were commenting on here, but I see now that those were just in this patch. I got confused by the "View" option on GitHub (which shows the new file but without the diff markings).

mravn-google reviewed (approved):
LGTM

brianegan committed:
More formatting corrections

mravn-google commented:
Excellent. Thanks, I'll merge once Travis is ready.

brianegan commented:
Just wanted to make sure this comment didn't get lost:

Totally understand these nitpicks, but I wonder about the style of this file. It has many lines like this:

https://github.com/flutter/flutter/pull/13820/files#diff-ecef29ade55b943f96ea50b77461443fR88

It also doesn't seem to follow the `dartfmt` style guide as many lines are over 80 characters and therefore, when I run `dartfmt`, I end up making changes to most of the file (so I've avoided doing that to keep the PR focused). Should I file a follow-up issue to fix the styling for this file overall?

brianegan commented:
Thanks again for all the help / good reviews getting this in the right place :D

mravn-google commented:
Yes, the Flutter repo is full of files that disagree with `dartfmt`. I do think the lines in question are true to the Flutter style guide though, both before and after your latest changes. But since writing the original versions of these lines, I've come to realize that one can often have the best of both worlds with enough line breaks and trailing commas.
